### PR TITLE
feat(custombackend): Change the Searcher to use a searchable index instead of concrete implementation

### PR DESCRIPTION
**Pull request by robertmogos**
…

### CODE REVIEW

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchResultsHandler.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+public interface SearchResultsHandler<Results> {
+
+    /***
+     * Called when the request has completed, either successfully or failing.
+     * <p>
+     * NOTE: One and only one of either <code>content</code> or <code>error</code> is guaranteed to be not null.
+     * </p>
+     *
+     * @param content
+     * @param error
```
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchResultsHandler.java]: would add a succinct description of each params to improve DX

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Map;
+
+/***
+ * Base class used to implement a custom backend with Algolia
+ * You need to make sure to implement the following method:
+ * {@link SearchTransformer#search(Object, SearchResultsHandler)}
+ * {@link SearchTransformer#map(Query)}
+ * {@link SearchTransformer#map(Object)}
+ */
+public class SearchTransformer<Parameters, Results> extends Searchable implements Transformable<Parameters, Results> {
```
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java]: Can we make this `abstract` to make it clear one needs to subclass it?

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchResultsHandler.java (hunk 1 above)
Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchResultsHandler.java]: There are several other cases of undocumented parameters, I won't comment those but the same applies 🙃

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    public Parameters map(@NonNull Query query) {
+        throw new UnsupportedOperationException("Not needed for now");
+    }
+
+    @Override
+    public Parameters map(@NonNull Query query, @NonNull Collection<String> disjunctiveFacets, @NonNull Map<String, ? extends Collection<String>> refinements) {
+        throw new UnsupportedOperationException("Not needed for now");
```
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java]: Not sure this error message is clear in the context a customer would sees it. Can it be more explicit, like `this method was not implemented yet; override it if you need it`?

Comment (PLNech) [reply instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java]: Same applies below

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+import org.json.JSONObject;
+
+import java.util.Collection;
+import java.util.Map;
+
+public interface Transformable<Parameters, Results> {
+
+    /***
+     * Search operation
```
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java]: Would describe what it does, like `Performs a search for the given Parameters, returning Results`

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+     *
+     * @param query
+     * @param disjunctiveFacets
+     * @param refinements
+     * @return
+     */
+    Parameters map(@NonNull Query query, @NonNull final Collection<String> disjunctiveFacets, @NonNull final Map<String, ? extends Collection<String>> refinements);
+
+    /***
+     * Transforms the custom backend result to an Algolia result.
```
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java]: results

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+     *
+     * @param results
+     * @return
+     */
+    JSONObject map(@NonNull Results results);
+
+
+    /***
+     * Search for facet value operation
```
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java]: Would explicit as well, like `Performs a search in the values for a given facet.`

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java [hunk 7]
```diff
@@ -889,7 +895,7 @@ public Searcher setQuery(@NonNull Query query) {
      * @return the Searcher's index.
      */
     @SuppressWarnings({"WeakerAccess", "unused"}) // For library users
-    public Index getIndex() {
+    public Searchable getIndex() {
```
[the remark this answers is not in the document]
Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: Done

File: instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java [hunk 8]
```diff
@@ -165,8 +167,8 @@ public static Searcher create(@NonNull final String appId, @NonNull final String
      * @return the new instance.
      */
     @SuppressWarnings({"WeakerAccess", "unused"}) // For library users
-    public static Searcher create(@NonNull final Index index) {
-        final String key = index.getRawIndexName();
+    public static Searcher create(@NonNull final Searchable index) {
```
[the remark this answers is not in the document]
Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/helpers/Searcher.java]: Done

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java (hunk 2 above)
Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java]: Done

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java (hunk 3 above)
Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/transformer/SearchTransformer.java]: Done

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java (hunk 6 above)
Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java]: Done

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java (hunk 5 above)
Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java]: Done

File: instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java (hunk 4 above)
Comment (robertmogos) [reply instantsearch/src/main/java/com/algolia/instantsearch/transformer/Transformable.java]: Done

**Comment by robertmogos** [comment]
@PLNech I re-added the method and deprecate it

**Pull request merged**